### PR TITLE
video: add a constructor with API preference parameter for VideoWriter

### LINKED ISSUE #8859: Unable to use gstreamer videowriter
System information (version)

OpenCV => 3.2
Operating System / Platform => Mac OS 10.12.5
Compiler => g++
Wrapper => Java/C++

It is impossible to instantiate a cvCreateVideoWriter_GStreamer because TRY_OPEN detects that a writer has already been created by ffmpeg (or possibly other cap).    FFMpeg jsut creates a file with the name of my gstreamer pipeline construction..

My call to VideoWriter():
writer.open("appsrc ! autovideoconvert ! vtenc_h264 ! h264parse ! mpegtsmux ! rtpmp2tpay ! udpsink host=127.0.0.1 port=5000", fourcc, capture.get(Videoio.CAP_PROP_FPS), frameSize);

I can temporarily fix the issue by moing the HAVE_GSTREAMER check above ffmpeg, but I supose the real fix would be to have ffmpeg fail if it thinks that it should not be a valid file.

### PR DESCRIPTION
**Pull request by sovrasov**
resolves #8859

### CODE REVIEW

File: modules/videoio/include/opencv2/videoio/videoio_c.h [hunk 1]
```diff
@@ -560,6 +560,14 @@ CVAPI(CvVideoWriter*) cvCreateVideoWriter( const char* filename, int fourcc,
                                            double fps, CvSize frame_size,
                                            int is_color CV_DEFAULT(1));
 
+/**
+
+*/
+CVAPI(CvVideoWriter*) cvCreateVideoWriterWithPreference(int api, const char* filename, int fourcc,
+                                           double fps, CvSize frame_size,
+                                           int is_color CV_DEFAULT(1));
```
Comment (alalek) [inline modules/videoio/include/opencv2/videoio/videoio_c.h]: Probably we should not add new exported `cv*` functions. Please make it local "static".

Comment (sovrasov) [inline modules/videoio/include/opencv2/videoio/videoio_c.h]: I've moved it to the `precomp.hpp` where other similar functions are declared.

File: modules/videoio/src/cap.cpp [hunk 2]
```diff
@@ -371,45 +371,56 @@ CV_IMPL CvCapture * cvCreateFileCapture (const char * filename)
...
+{
     CV_UNUSED(frameSize);
     CV_UNUSED(is_color);
 
-    //CV_FUNCNAME( "cvCreateVideoWriter" );
-
     CvVideoWriter *result = 0;
 
     if(!fourcc || !fps)
         TRY_OPEN(result, cvCreateVideoWriter_Images(filename))
```
Comment (mshabunin) [inline modules/videoio/src/cap.cpp]: Maybe we could define additional API IDs for built-in MotionJPEG and Images backends? Or maybe we should add new set of IDs for _VideoWriter_ specifically: _WRI_ANY_, _WRI_FFMPEG_, etc.? What do you think?

File: modules/videoio/src/cap.cpp [hunk 3]
```diff
@@ -371,45 +371,56 @@ CV_IMPL CvCapture * cvCreateFileCapture (const char * filename)
...
-
-#ifdef HAVE_GSTREAMER
-    TRY_OPEN(result, cvCreateVideoWriter_GStreamer(filename, fourcc, fps, frameSize, is_color))
-#endif
+    switch(apiPreference)
+    {
+        default:
+            //exit if the specified API is unavaliable
+            if (apiPreference != CV_CAP_ANY) break;
+        case(CV_CAP_FFMPEG):
```
Comment (alalek) [inline modules/videoio/src/cap.cpp]: BTW, Why "case" values are in brackets?

**Comment by alalek** [comment]
Looks good to me. Well done! :+1:

File: modules/videoio/src/cap.cpp (hunk 3 above)
Comment (mshabunin) [inline modules/videoio/src/cap.cpp]: In my opinion wrapping whole _case_ blocks into _#ifdef..#endif_ will look more clean:
```.cpp
#ifdef HAVE_FFMPEG
        case CV_CAP_FFMPEG:
            TRY_OPEN(result, cvCreateVideoWriter_FFMPEG_proxy (filename, fourcc, fps, frameSize, is_color))
            if (apiPreference != CV_CAP_ANY) break;
#endif

#ifdef HAVE_VFW
        case CV_CAP_VFW:
            TRY_OPEN(result, cvCreateVideoWriter_VFW(filename, fourcc, fps, frameSize, is_color))
            if (apiPreference != CV_CAP_ANY) break;
#endif
```

Comment (sovrasov) [inline modules/videoio/src/cap.cpp]: Don't know, just a habit.

File: modules/videoio/src/cap.cpp (hunk 2 above)
Comment (sovrasov) [inline modules/videoio/src/cap.cpp]: I've defined `CAP_OCV_MJPEG` and we already have `CAP_IMAGES`.

**Comment by mshabunin** [comment]
:+1:

**Pull request merged**